ldm/dream/args.py

```diff
@@ -0,0 +1,485 @@
+
+  # The args object acts like a namespace object
+  print(a.model)
+
+You can set attributes in the usual way, use vars(), etc.:
+
+  a.model = 'something-else'
+  do_something(**vars(a))
+
+It is helpful in saving metadata:
+
+  # To get a json representation of all the values, allowing
+  # you to override any values dynamically
+  j = a.json(seed=42)
+
+  # To get the prompt string with the switches, allowing you
+  # to override any values dynamically
+  j = a.prompt_str(seed=42)
+
+If you want to access the namespace objects from the shell args or the
+parsed command directly, you may use the values returned from the
+original calls to parse_args() and parse_cmd(), or get them later
+using the _arg_switches and _cmd_switches attributes. This can be
+useful if both the args and the command contain the same attribute and
+you wish to apply logic as to which one to use. For example:
+
+  a = Args()
+  args    = a.parse_args()
+  opts    = a.parse_cmd(string)
+  do_grid = args.grid or opts.grid
+
+To add new attributes, edit the _create_command_arg_parser() and
+_create_cmd_parser() methods.
+
+We also export the function build_metadata
+"""
+
+import argparse
+import shlex
+import json
+
+SAMPLER_CHOICES = [
+    'ddim',
+    'k_dpm_2_a',
+    'k_dpm_2',
+    'k_euler_a',
+    'k_euler',
+    'k_heun',
+    'k_lms',
+    'plms',
+]
+
+# is there a way to pick this up during git commits?
+APP_ID      = 'lstein/stable-diffusion'
+APP_VERSION = 'v1.15'
+
+class Args(object):
+    def __init__(self):
+        '''Initialize new Args class. Takes no arguments'''
+        self._arg_parser   = self._create_arg_parser()
+        self._cmd_parser   = self._create_cmd_parser()
+        self._arg_switches = self.parse_args()   # so there's something there
+        self._cmd_switches = self.parse_cmd('')  # so there's something there
+
+    def parse_args(self):
+        '''Parse the shell switches and store.'''
+        self._arg_switches = self._arg_parser.parse_args()
+        return self._arg_switches
+
+    def parse_cmd(self,cmd_string):
+        '''Parse a dream>-style command string '''
+        command = cmd_string.replace("'", "\\'")
+        try:
+            elements = shlex.split(command)
+        except ValueError:
+            print(traceback.format_exc(), file=sys.stderr)
+            return
+        switches = ['']
+        switches_started = False
+
+        for el in elements:
+            if el[0] == '-' and not switches_started:
+                switches_started = True
+            if switches_started:
+                switches.append(el)
+            else:
+                switches[0] += el
+                switches[0] += ' '
+        switches[0] = switches[0][: len(switches[0]) - 1]
+        self._cmd_switches = self._cmd_parser.parse_args(switches)
+        return self._cmd_switches
+
+    def json(self,**kwargs):
+        return json.dumps(self.to_dict(**kwargs))
+
+    def to_dict(self,**kwargs):
+        a = vars(self)
```

Comment:
A function like this is the *only* place I don't mind seeing non-descriptive names - it's incredibly self-evident, and more importantly - this function is exceedingly unlikely to *grow in the future* (at least not by very much).